garage/tf/algos/batch_polopt.py

```diff
@@ -64,6 +75,112 @@ def log_diagnostics(self, paths):
+            ]
+
+        for idx, path in enumerate(paths):
+            path_baselines = np.append(all_path_baselines[idx], 0)
+            deltas = (path['rewards'] + self.discount * path_baselines[1:] -
+                      path_baselines[:-1])
+            path['advantages'] = special.discount_cumsum(
+                deltas, self.discount * self.gae_lambda)
+            path['deltas'] = deltas
+
+        for idx, path in enumerate(paths):
+            # baselines
+            path['baselines'] = all_path_baselines[idx]
+            baselines.append(path['baselines'])
+
+            # returns
+            path['returns'] = special.discount_cumsum(path['rewards'],
+                                                      self.discount)
+            returns.append(path['returns'])
+
+        # make all paths the same length
+        obs = [path['observations'] for path in paths]
+        obs = tensor_utils.pad_tensor_n(obs, max_path_length)
+
+        actions = [path['actions'] for path in paths]
+        actions = tensor_utils.pad_tensor_n(actions, max_path_length)
+
+        rewards = [path['rewards'] for path in paths]
+        rewards = tensor_utils.pad_tensor_n(rewards, max_path_length)
+
+        returns = [path['returns'] for path in paths]
+        returns = tensor_utils.pad_tensor_n(returns, max_path_length)
+
+        baselines = tensor_utils.pad_tensor_n(baselines, max_path_length)
+
+        agent_infos = [path['agent_infos'] for path in paths]
+        agent_infos = tensor_utils.stack_tensor_dict_list([
+            tensor_utils.pad_tensor_dict(p, max_path_length)
+            for p in agent_infos
+        ])
+
+        env_infos = [path['env_infos'] for path in paths]
+        env_infos = tensor_utils.stack_tensor_dict_list([
+            tensor_utils.pad_tensor_dict(p, max_path_length) for p in env_infos
+        ])
+
+        valids = [np.ones_like(path['returns']) for path in paths]
+        valids = tensor_utils.pad_tensor_n(valids, max_path_length)
+
+        average_discounted_return = (np.mean(
+            [path['returns'][0] for path in paths]))
+
+        undiscounted_returns = [sum(path['rewards']) for path in paths]
+        self.episode_reward_mean.extend(undiscounted_returns)
+
+        ent = np.sum(self.policy.distribution.entropy(agent_infos) *
+                     valids) / np.sum(valids)
+
+        samples_data = dict(
+            observations=obs,
+            actions=actions,
+            rewards=rewards,
+            baselines=baselines,
+            returns=returns,
+            valids=valids,
+            agent_infos=agent_infos,
+            env_infos=env_infos,
+            paths=paths,
+            average_return=np.mean(undiscounted_returns),
+        )
+
+        tabular.record('Iteration', itr)
+        tabular.record('AverageDiscountedReturn', average_discounted_return)
+        tabular.record('AverageReturn', np.mean(undiscounted_returns))
```

Comment:
couldn't it get most of this by calling `super().process_samples(itr, paths)` first?

Reply:
It might be more involved than expected, because several fields are produced by `tensor_utils`, but the np version doesn't.

Reply:
One thing we can try is 
- make a full version of np `process_samples()` (compute and log everything).
- inherit tf `BatchPolopt` from np `BatchPolopt`.
- call `super().process_samples(itr, paths)` to get processed sample.
- cast processed sample fields from ndarray to tensor.

Reply:
Do you want me to implement this way? I think we are trying to separate np and tf algorithms. Put the full version of process_samples() to np seems confusing.

Reply:
i agree -- they should be separate.

perhaps it would be worth it to create a shared helper function for the most common parts of process samples (e.g. calculating common stats like StdReturn, Entropy, etc.)?

Reply:
I don't come out a good idea of this. Since I remember you said you are against the utility things. And it seems not worth here to me. @naeioi What do you think?

Reply:
let's fix it in a different PR